### PR TITLE
tfdsv4 fix carla configs

### PR DESCRIPTION
**Pull request by lcadalzo**
Note that the carla attacks don't currently work with the new tfdsv4 pipeline. This is because the label keys now have a batch dimension, e.g. the old format that works with the attacks is
```
(Pdb) y_target[0]['boxes'].shape
(10, 4)
```
and the new:
```
(Pdb) y_target[0]['boxes'].shape
(1, 10, 4)
```

I'm not intending to address that in this PR, so I tested with `--skip-attack`.

### CODE REVIEW

**Comment by lcadalzo** [comment]
marking WIP since I still need to add the dataset's .py file

**Comment by lcadalzo** [comment]
the only change between `carla_obj_det_dev.py` added here and that on `develop` in `armory.data` is in `_split_generators()` and is functionally equivalent

File: armory/datasets/adversarial/carla_obj_det_dev/carla_obj_det_dev.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
+"""carla_obj_det_dev dataset."""
+
+import collections
+import json
+import os
+from copy import deepcopy
+import numpy as np
+
+import tensorflow.compat.v1 as tf
```
[the remark this answers is not in the document]
Comment (lcadalzo) [reply armory/datasets/adversarial/carla_obj_det_dev/carla_obj_det_dev.py:9]: that's from a copy paste, but let me see if I can get rid of that

[the remark this answers is not in the document]
Comment (lcadalzo) [reply armory/datasets/adversarial/carla_obj_det_dev/carla_obj_det_dev.py]: done

**Comment by lcadalzo** [comment]
ready for re-review @davidslater. Some of the preprocessing code this PR touches is being modified in #1797, but I think we can merge this in and make sure to test carla and xview when when reviewing the latter PR

**Comment by lcadalzo** [comment]
@davidslater ready to go. The commits from today are just renaming a function and removing code that got duplicated in the handling of the merge conflict

**Comment by davidslater** [comment]
The error may be due to changes in develop that happened after we branched to tfdsv4

**Comment by lcadalzo** [comment]
meaning there have been fixes in `develop` that haven't made their way into `tfdsv4` branch yet?

**Comment by davidslater** [comment]
No, on second look, those occurred before develop was merged into tfdsv4, so I don't know.

**Comment by lcadalzo** [comment]
ah ok, the issue is occurring because similar to [this](https://github.com/twosixlabs/armory/issues/1814) issue, the tfdsv4 pipeline outputs labels with a batch dimension e.g.

```
(Pdb) y['boxes'].shape
(1, 15, 4)
```

whereas the code is expecting `(15, 4)`.

**Comment by lcadalzo** [comment]
The reason mAP _doesn't_ fail with a similar error is because of [these](https://github.com/twosixlabs/armory/blob/v0.16.1/armory/metrics/task.py#L650-L651) lines. I think this PR should be fine to merge, and we'll need to add a separate issue to tweak tide calculation to be compatible with this form, since the tf preprocessing will always output with a batch dimension

File: armory/datasets/adversarial/carla_obj_det_dev/carla_obj_det_dev.py [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+  title = { {CARLA}: {An} Open Urban Driving Simulator},
+  author = {Alexey Dosovitskiy and German Ros and Felipe Codevilla and Antonio Lopez and Vladlen Koltun},
+  booktitle = {Proceedings of the 1st Annual Conference on Robot Learning},
+  pages = {1--16},
+  year = {2017}
+}
+"""
+
+# fmt: off
+_URLS = "https://armory-public-data.s3.us-east-2.amazonaws.com/carla/carla_od_dev_2.0.0.tar.gz"
```
Comment (jprokos26) [inline armory/datasets/adversarial/carla_obj_det_dev/carla_obj_det_dev.py:27]: Should coincide with `tfdsv4` convention.
```suggestion
_URL = "https://armory-public-data.s3.us-east-2.amazonaws.com/carla/carla_od_dev_2.0.0.tar.gz"
```
See https://github.com/twosixlabs/armory/blob/tfdsv4/armory/datasets/adversarial/carla_over_obj_det_dev/carla_over_obj_det_dev.py#L26.

File: armory/datasets/adversarial/carla_obj_det_dev/carla_obj_det_dev.py [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+        return tfds.core.DatasetInfo(
+            builder=self,
+            description=_DESCRIPTION,
+            features=tfds.features.FeaturesDict(features),
+            citation=_CITATION,
+        )
+
+    def _split_generators(self, dl_manager: tfds.download.DownloadManager):
+        """Returns SplitGenerators."""
+        path = dl_manager.download_and_extract(_URLS)
```
Comment (jprokos26) [inline armory/datasets/adversarial/carla_obj_det_dev/carla_obj_det_dev.py:104]:
```suggestion
        path = dl_manager.download_and_extract(_URL)
```

**Comment by jprokos26** [comment]
Throwing error when running `carla_obj_det_adversarialpatch_undefended.json` but [url](https://armory-public-data.s3.us-east-2.amazonaws.com/carla/carla_od_dev_2.0.0.tar.gz) provided is valid, was the cached dataset uploaded as well?

```cmd
  File "/workspace/armory/datasets/download.py", line 70, in download
    download_file_from_s3(str(key), str(filepath), public=public)
    │                         │         │                 └ True
    │                         │         └ PosixPath('/armory/datasets/new_builds/cache/carla_obj_det_dev__2.0.0__cache.tar.gz')
    │                         └ 'datasets/cache/carla_obj_det_dev__2.0.0__cache.tar.gz'
    └ <function download_file_from_s3 at 0x7f8c2f2b2c10>

  File "/workspace/armory/datasets/download.py", line 38, in download_file_from_s3
    raise KeyError(f"File {key} not available in {bucket} bucket.")

KeyError: 'File datasets/cache/carla_obj_det_dev__2.0.0__cache.tar.gz not available in armory-public-data bucket.'
```

**Review by jprokos26 — CHANGES_REQUESTED** [review]
Convention update.

**Comment by lcadalzo** [comment]
> Throwing error when running carla_obj_det_adversarialpatch_undefended.json but [url](https://armory-public-data.s3.us-east-2.amazonaws.com/carla/carla_od_dev_2.0.0.tar.gz) provided is valid, was the cached dataset uploaded as well?

@jprokos26 try again, it looks like upload to `armory-public-data/datasets/cache` hadn't occurred, but I just uploaded

**Comment by jprokos26** [comment]
Now throwing an error at end of run using `--skip-attack`:

```python
File "/workspace/armory/metrics/task.py", line 940, in object_detection_mAP_tide
tidecv_ground_truth_list = armory_to_tide(y, image_id, is_detection=False)
                           │              │  └ array([69734696, 69734696, 69734696, 69734696, 69734696, 69734696,
                           │              │           69734696, 69734696, 69734696, 69734696, 69734696, 6...
                           │              └ {'area': array([[1869, 3543,  120,  107,  145,  102,  116,  130,  139,  149,  122,
                           │                         154, 3893, 7951,  172]]), 'boxes'...
                           └ <function armory_to_tide at 0x7fb2c23231f0>
File "/workspace/armory/metrics/task.py", line 882, in armory_to_tide
x1, y1, x2, y2 = y["boxes"]
                 └ {'area': array([1869, 3543,  120,  107,  145,  102,  116,  130,  139,  149,  122,
                           154, 3893, 7951,  172]), 'boxes': a...
ValueError: too many values to unpack (expected 4)
```

**Comment by lcadalzo** [comment]
> Now throwing an error at end of run using --skip-attack:

@jprokos26 see [this](https://github.com/twosixlabs/armory/pull/1788#issuecomment-1340067742) comment and the ones above it. I still think there's a decision to be made here that applies to various datasets: do we modify scenario code to remove the batch dimension, or do we update metrics to expect the batch dimension. I'm deferring on this decision for now and not including a fix for that in this PR (we're tracking via #1814)

**Comment by jprokos26** [comment]
> > Now throwing an error at end of run using --skip-attack:
> 
> @jprokos26 see [this](https://github.com/twosixlabs/armory/pull/1788#issuecomment-1340067742) comment and the ones above it. I still think there's a decision to be made here that applies to various datasets: do we modify scenario code to remove the batch dimension, or do we update metrics to expect the batch dimension. I'm deferring on this decision for now and not including a fix for that in this PR (we're tracking via #1814)

Gotcha, then I agree this PR is ready to be merged if that issue is being tracked as the dataset is loading as expected now.

**Review by jprokos26 — APPROVED** [review]

**Pull request merged**